###  Loading Data & Libraries

In [162]:
import pandas as pd
import numpy as np

In [163]:
data = pd.read_csv("airlines.csv")

### Show Sample Of Data

In [164]:
data.head()

Unnamed: 0    id        day      airline        destination    dest_region  \
0           0  1351    Tuesday  UNITED INTL             KANSAI           Asia   
1           1   373     Friday       ALASKA  SAN JOSE DEL CABO  Canada/Mexico   
2           2  2820   Thursday        DELTA        LOS ANGELES        West US   
3           3  1157    Tuesday    SOUTHWEST        LOS ANGELES        West US   
4           4  2992  Wednesday     AMERICAN              MIAMI        East US   

  dest_size boarding_area   dept_time  wait_min     cleanliness  \
0       Hub  Gates 91-102  2018-12-31     115.0           Clean   
1     Small   Gates 50-59  2018-12-31     135.0           Clean   
2       Hub   Gates 40-48  2018-12-31      70.0         Average   
3       Hub   Gates 20-39  2018-12-31     190.0           Clean   
4       Hub   Gates 50-59  2018-12-31     559.0  Somewhat clean   

          safety        satisfaction  
0        Neutral      Very satisfied  
1      Very safe      Very satisfied  
2  Somewhat safe             Neutral  
3      Very safe  Somewhat satsified  
4      Very safe  Somewhat satsified

In [165]:
data.tail()

Unnamed: 0    id       day        airline   destination dest_region  \
2472        2804  1475   Tuesday         ALASKA  NEW YORK-JFK     East US   
2473        2805  2222  Thursday      SOUTHWEST       PHOENIX     West US   
2474        2806  2684    Friday         UNITED       ORLANDO     East US   
2475        2807  2549   Tuesday        JETBLUE    LONG BEACH     West US   
2476        2808  2162  Saturday  CHINA EASTERN       QINGDAO        Asia   

     dest_size boarding_area   dept_time  wait_min     cleanliness  \
2472       Hub   Gates 50-59  2018-12-31     280.0  Somewhat clean   
2473       Hub   Gates 20-39  2018-12-31     165.0           Clean   
2474       Hub   Gates 70-90  2018-12-31      92.0           Clean   
2475     Small    Gates 1-12  2018-12-31      95.0           Clean   
2476     Large    Gates 1-12  2018-12-31     220.0           Clean   

             safety        satisfaction  
2472        Neutral  Somewhat satsified  
2473      Very safe      Very satisfied  
2474      Very safe      Very satisfied  
2475  Somewhat safe      Very satisfied  
2476      Very safe  Somewhat satsified

In [166]:
data.shape

(2477, 13)

### Show More Info 

In [167]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2477 entries, 0 to 2476
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     2477 non-null   int64  
 1   id             2477 non-null   int64  
 2   day            2477 non-null   object 
 3   airline        2477 non-null   object 
 4   destination    2477 non-null   object 
 5   dest_region    2477 non-null   object 
 6   dest_size      2477 non-null   object 
 7   boarding_area  2477 non-null   object 
 8   dept_time      2477 non-null   object 
 9   wait_min       2477 non-null   float64
 10  cleanliness    2477 non-null   object 
 11  safety         2477 non-null   object 
 12  satisfaction   2477 non-null   object 
dtypes: float64(1), int64(2), object(10)
memory usage: 251.7+ KB


In [168]:
data.describe()

Unnamed: 0           id     wait_min
count  2477.000000  2477.000000  2477.000000
mean   1411.785224  1813.363746   166.084376
std     815.300508   945.226305   107.643625
min       0.000000     1.000000    15.000000
25%     709.000000  1038.000000   105.000000
50%    1414.000000  1859.000000   145.000000
75%    2122.000000  2581.000000   185.000000
max    2808.000000  9004.000000  1365.000000

In [169]:
data.duplicated().any()

False

### Above cells' conclusion

- Columns need some modifications
- Dataset has no missing values
- Dataset has no complete duplicates
- We remove `Unnamed: 0`
- We need to convert `dept_time` to datetime



# Cleaning

In [170]:
# remove Unnamed: 0 column 

data.drop(columns = ["Unnamed: 0"], inplace = True)  

In [171]:
# convert dept_time datatype to datetime

data["dept_time"] = pd.to_datetime(data["dept_time"], format= '%Y-%m-%d')  

###  Discovering `day` column

In [172]:
# show nuique value

data["day"].unique().tolist()

['Tuesday', 'Friday', 'Thursday', 'Wednesday', 'Saturday', 'Sunday', 'Monday']

In [173]:
# output value count of each unique value

data["day"].value_counts()

Tuesday      535
Wednesday    518
Thursday     488
Friday       368
Sunday       267
Saturday     210
Monday        91
Name: day, dtype: int64

###  After Discovering `Day` column 

- Traveling available over week days 
- No nan value found
- No other word found out column scope

#### The `most` common day people travel
- Day `Tuesday` 
- Count of people traveling is `535`

#### The `lowest`  common day people travel
- Day `Monday` 
- Count of people traveling is `91`

### Discovering `airline` column

In [174]:
# show nuique value

data["airline"].unique()

array(['UNITED INTL', 'ALASKA', 'DELTA', 'SOUTHWEST', 'AMERICAN',
       'JETBLUE', 'AEROMEXICO', 'AIR CANADA', 'UNITED', 'INTERJET',
       'TURKISH AIRLINES', 'AIR FRANCE/KLM', 'HAWAIIAN AIR', 'COPA',
       'WOW', 'KOREAN AIR', 'EMIRATES', 'AVIANCA', 'AER LINGUS',
       'CATHAY PACIFIC', 'BRITISH AIRWAYS', 'PHILIPPINE AIRLINES',
       'LUFTHANSA', 'QANTAS', 'FRONTIER', 'CHINA EASTERN', 'EVA AIR',
       'VIRGIN ATLANTIC', 'AIR NEW ZEALAND', 'SINGAPORE AIRLINES',
       'AIR CHINA', 'CHINA SOUTHERN', 'ANA ALL NIPPON'], dtype=object)

In [212]:
# output value count of each unique value

data["airline"].value_counts()

UNITED                 526
ALASKA                 315
SOUTHWEST              184
AMERICAN               175
DELTA                  170
UNITED INTL            155
JETBLUE                144
AIR CANADA             133
AIR FRANCE/KLM         120
CATHAY PACIFIC          81
LUFTHANSA               64
FRONTIER                41
EMIRATES                39
WOW                     38
VIRGIN ATLANTIC         32
TURKISH AIRLINES        30
QANTAS                  28
AIR NEW ZEALAND         21
BRITISH AIRWAYS         21
INTERJET                18
AER LINGUS              18
SINGAPORE AIRLINES      16
COPA                    16
PHILIPPINE AIRLINES     14
CHINA EASTERN           13
HAWAIIAN AIR            12
KOREAN AIR              12
AEROMEXICO              11
EVA AIR                 10
AIR CHINA                8
AVIANCA                  6
CHINA SOUTHERN           4
ANA ALL NIPPON           2
Name: airline, dtype: int64

###  After Discovering `airline` column 

- No nan value found
- We have `33` unique airline 

####  The `most` common trips people travel  

- airline `UNITED`
- Count of trip is `526`

#### The `lowest`  common trips people travel  
- airline `ANA ALL NIPPON`
- Count of trip is `2`

### Discovering `destination` column

In [176]:
# show nuique value

data["destination"].unique()

array(['KANSAI', 'SAN JOSE DEL CABO', 'LOS ANGELES', 'MIAMI', 'NEWARK',
       'LONG BEACH', 'MEXICO CITY', 'TORONTO', 'PORTLAND', 'SAN DIEGO',
       'BOSTON', 'SPOKANE', 'GUADALAJARA', 'MINNEAPOLIS-ST. PAUL',
       'NEW YORK-JFK', 'ISTANBUL', 'BALTIMORE', 'LAS VEGAS', 'SHANGHAI',
       'TOKYO-NARITA', 'PARIS-DE GAULLE', 'HONOLULU', 'DALLAS-FT. WORTH',
       'PANAMA CITY', 'PHOENIX', 'REYKJAVIK', 'SAN ANTONIO', 'HONG KONG',
       'SEOUL', 'DUBAI', "CHICAGO-O'HARE", 'INDIANAPOLIS', 'SAN SALVADOR',
       'SALT LAKE CITY', 'BEIJING', 'DUBLIN', 'WASHINGTON DC-DULLES',
       'LONDON HEATHROW', 'MANILA', 'RALEIGH-DURHAM', 'VANCOUVER',
       'MUNICH', 'NEW ORLEANS', 'FRANKFURT', 'SYDNEY', 'KAHULUI',
       'AMSTERDAM', 'ATLANTA', 'SEATTLE', 'DETROIT', 'SANTA BARBARA',
       'PHILADELPHIA', 'DENVER', 'BAKERSFIELD', 'AUSTIN', 'CALGARY',
       'TAIPEI', 'ONTARIO (CALIF)', 'BURBANK', 'CHARLOTTE', 'AUCKLAND',
       'SINGAPORE', 'ORLANDO', 'NASHVILLE', 'WUHAN', 'HOUSTON-BUSH',
       'FT

In [177]:
# output value count of each unique value

data["destination"].value_counts()

LOS ANGELES        185
PARIS-DE GAULLE    126
DENVER              96
HONG KONG           95
LONDON HEATHROW     91
                  ... 
WUHAN                4
ONTARIO (CALIF)      3
NEW ORLEANS          3
QINGDAO              3
EUGENE               1
Name: destination, Length: 72, dtype: int64

###  After Discovering `destination` column 

- No nan value found
- We have `72` unique airline

####  The `most` common trips people traveling To

- airline `LOS ANGELES`
- Count of trip is `185`

#### The `lowest`  common trips people traveling To  
- airline `EUGENE`
- Count of trip is `2`

In [178]:
# Ensure that the airline and destination airport not same

assert data[data["airline"] == data["destination"]].shape[0] == 0  # if yes not return any else return AssertionError

### Discovering `dest_region` column

In [179]:
# show nuique value

data["dest_region"].unique()

array(['Asia', 'Canada/Mexico', 'West US', 'East US', 'Midwest US',
       'EAST US', 'Middle East', 'Europe', 'eur', 'Central/South America',
       'Australia/New Zealand', 'middle east'], dtype=object)

In [180]:
# output value count of each unique value

data["dest_region"].value_counts()

West US                  864
East US                  367
Europe                   272
Midwest US               251
Asia                     226
Canada/Mexico            198
eur                       79
EAST US                   69
Australia/New Zealand     60
Middle East               48
Central/South America     22
middle east               21
Name: dest_region, dtype: int64

###  Column has some error


- We have same words with different format
- We have some words and shortcut of this words 
- We need to fixed this error


In [181]:
# fixed dest_region column

# 1- convert all values to upper case
data["dest_region"] = data["dest_region"].str.upper()

# 2- replace EUR with EUROPE 
data.loc[data["dest_region"] == "EUR", "dest_region"] = "EUROPE"

In [182]:
# show nuique value after modifying

data["dest_region"].unique()

array(['ASIA', 'CANADA/MEXICO', 'WEST US', 'EAST US', 'MIDWEST US',
       'MIDDLE EAST', 'EUROPE', 'CENTRAL/SOUTH AMERICA',
       'AUSTRALIA/NEW ZEALAND'], dtype=object)

In [183]:
# output value count of each unique value after modifying

data["dest_region"].value_counts()

WEST US                  864
EAST US                  436
EUROPE                   351
MIDWEST US               251
ASIA                     226
CANADA/MEXICO            198
MIDDLE EAST               69
AUSTRALIA/NEW ZEALAND     60
CENTRAL/SOUTH AMERICA     22
Name: dest_region, dtype: int64

###  After Discovering `dest_region` column 

- No nan value found
- We have `9` regions

#### The `most` common region

- Region name is `WEST US`
- Number of trips is `864`

#### The `lowest` common region

- Region name is `CENTRAL/SOUTH AMERICA`
- Number of trips is `22`



### Discovering `dest_size` column

In [184]:
# show unique values 

data["dest_size"].unique()

array(['Hub', 'Small', '    Hub', 'Medium', 'Large', 'Hub     ',
       '    Small', 'Medium     ', '    Medium', 'Small     ',
       '    Large', 'Large     '], dtype=object)

###  Column has some error

- We have same words with different format and extra space
- We need to fixed this error

In [185]:
# fixed dest_size column

# remove extra space
data["dest_size"] = data["dest_size"].str.strip()

In [186]:
# show nuique value after modifying

data["dest_size"].unique()

array(['Hub', 'Small', 'Medium', 'Large'], dtype=object)

In [187]:
# output value count of each unique value 

data["dest_size"].value_counts()

Hub       1550
Medium     600
Small      200
Large      127
Name: dest_size, dtype: int64

###  After Discovering `dest_size` column 

- No nan value found
- We have `4` destance

#### The `most` common destance

- Destance name `Hub`
- Number of trips is `1550`

#### The `lowest` common destance

- Destance name `Large`
- Number of trips is `127`


### Discovering `boarding_area` column

In [188]:
data["boarding_area"].unique()

array(['Gates 91-102', 'Gates 50-59', 'Gates 40-48', 'Gates 20-39',
       'Gates 1-12', 'Gates 70-90', 'Gates 60-69'], dtype=object)

In [189]:
data["boarding_area"].value_counts()

Gates 1-12      609
Gates 91-102    474
Gates 50-59     467
Gates 70-90     383
Gates 20-39     225
Gates 40-48     176
Gates 60-69     143
Name: boarding_area, dtype: int64

###  After Discovering `boarding_area` column 

- No nan value found
- We have `6` Gates
- Column no need any modification


### Discovering `wait_min` column

In [190]:
data["wait_min"].describe()

count    2477.000000
mean      166.084376
std       107.643625
min        15.000000
25%       105.000000
50%       145.000000
75%       185.000000
max      1365.000000
Name: wait_min, dtype: float64

### Column has some error
- Column has outliers
- We need to fixed this error

In [213]:
# Calculating outliers

Q1,Q2,Q3 = 105.000000, 145.000000, 185.000000
IQR = Q3 - Q1
upper = Q3 + (1.5 * IQR)
lower = Q1 - (1.5 * IQR)

# replace outliers values with nan
data.loc[(data["wait_min"] > upper) | (data["wait_min"] < lower), "wait_min"] = np.nan

In [214]:
# see how meny row repleced with nan value

data["wait_min"].isna().sum()

155

In [249]:
# fill nan value with mean of wait_min of each dset_region
    
# get mean of each group in dest_region column
mean_value = data.groupby(["dest_region"])["wait_min"].mean()

# set dest_region as index   
data = data.set_index(["dest_region"])

# fill nan value of each group with mean of wait_min column of this group
data['wait_min'] = data['wait_min'].fillna(mean_value)

# convert wait_min to column again
data = data.reset_index()

In [256]:
# convetr wait_min to integer

data["wait_min"] = data["wait_min"].astype("int")

### Discovering `cleanliness` column

In [195]:
# show unique values

data["cleanliness"].unique()

array(['Clean', 'Average', 'Somewhat clean', 'Somewhat dirty', 'Dirty'],
      dtype=object)

In [196]:
# output value counts

data["cleanliness"].value_counts()

Somewhat clean    1175
Clean              885
Average            389
Somewhat dirty      26
Dirty                2
Name: cleanliness, dtype: int64

### column need some modification

- Somewhat clean, Somewhat dirty equal to Average 
- We need to replace this two word to Average


In [197]:
# fixed cleanliness column

data.loc[(data["cleanliness"] == "Somewhat clean") | (data["cleanliness"] == "Somewhat dirty"), "cleanliness"] = "Average"

In [198]:
# show unique values after modifying

data["cleanliness"].value_counts()

Average    1590
Clean       885
Dirty         2
Name: cleanliness, dtype: int64

### After Discovering `cleanliness` column

- most people see trip not pure clean, dirty  (Average)


### Discovering `safety` column

In [194]:
data["safety"].unique()

array(['Neutral', 'Very safe', 'Somewhat safe', 'Very unsafe',
       'Somewhat unsafe'], dtype=object)

In [200]:
data["safety"].value_counts()

Very safe          1296
Somewhat safe       864
Neutral             303
Very unsafe           8
Somewhat unsafe       6
Name: safety, dtype: int64

###  After Discovering `safety` column 

- Most people see trip `Very safe` 


### Discovering `satisfaction` column

In [202]:
data["satisfaction"].unique()

array(['Very satisfied', 'Neutral', 'Somewhat satsified',
       'Somewhat unsatisfied', 'Very unsatisfied'], dtype=object)

In [204]:
data["satisfaction"].value_counts()

Somewhat satsified      1349
Very satisfied           596
Neutral                  486
Somewhat unsatisfied      42
Very unsatisfied           4
Name: satisfaction, dtype: int64

###  After Discovering `satisfaction` column 

- Most people see trip `Somewhat satsified` 
